### DISCUSSION TITLE
Some questions about using Jason as a library

### DISCUSSION THREAD

**Discussion by ChrisL20J** [jason-lang/jason#113] [Q&A] [answered]
Hi there,

I am currently working on a project that uses Jason as a library, and I'm trying to manage multiple agents within a MAS. However, I’m facing several challenges.

1. Adding Beliefs using `createAgent()`:
I am struggling to add beliefs when using the createAgent() method under jason.runtime.RuntimeServices. According to the [documentation](https://jason-lang.github.io/api/jason/runtime/RuntimeServices.html#createAgent(java.lang.String,java.lang.String,java.lang.String,java.util.List,jason.mas2j.ClassParameters,jason.runtime.Settings,jason.asSemantics.Agent)), there is a parameter to do this, but I’m unsure how to define ClassParameters to add a new belief. Could someone provide an example of how this can be done?
![image](https://github.com/user-attachments/assets/654d58a9-ce4d-497f-932b-a6e40757a668)
```
Creates a new agent with agName from source agSource, using agClass as agent class (default value is "jason.asSemantics.Agent"), archClasses as agent architecture classes, bbPars as the belief base class (default value is "DefaultBeliefBase"), stts as Settings (default value is new Settings()), and father is the agent creating this agent (null is none). if no archClasses is informed (null value), if fathers is informed use father's ag archs else use default ag archs (see registerDefaultAgArch)
Example: createAgent("bob", "bob.asl", "mypkg.MyAgent", null, null, null); Returns the name of the agent
```

2.Retrieving Agent Beliefs from Java:
Is there any method to actively retrieve the belief of an activated agent from the Java side? I know how to get the list of current agents managed by MAS, but I couldn’t find a method like getAgent() or something similar. Is there any way to get or load the current information (not just the initial agent/ASL code) of an agent?

3.Backing Up and Resuming Agents on Another MAS:
Is there a method to back up the current status of agents and resume their work in another MAS environment?

4.Loading java method and calling it from an agent (.asl file):
The agent can call the Java method from the `Env.java` file, which I had designated in the mas2j file. However, it couldn't call any method that is not in the `Env.java` file from the agent using the same method shown in [this demo](https://github.com/jason-lang/jason/tree/master/demos/create-agent). 
Considering `Env.java` is located under `src/main/java/com/xxx/` and other java classes are located under `src/main/java/com/xxx/mylib/`.
But I still got an error like this
` warning: The internal action class for 'mylib.method(90)' was not loaded! Error: java.lang.ClassNotFoundException: mylib.method`.
Is it a method that cannot be used while using Jason as a library?

I’ve been developing this on my own for months and have struggled to find information that can help me resolve these issues. I would greatly appreciate any assistance or guidance on any of these problems. Thank you.

**Comment by jomifred** [comment, marked as answer]
Hello Chris,

indeed, to use Jason as a library, a lot of knowledge about the API is required. 

1. the createAgent is not meant to be used to add beliefs. Even the BB parameter is used to inform a java class that implements a  custom BB. It simply creates the agent. From the [RuntimeServices ](https://jason-lang.github.io/api/jason/runtime/RuntimeServices.html) API, you can use `getAgentSnapshot` to get a copy of the agent, in an [Agent ](https://jason-lang.github.io/api/jason/asSemantics/Agent.html)instance.  From this class you can inspect all BB of the agent. Notice that it is a copy of the agent, not the agent itself.
2. From class [RunLocalMAS ](https://jason-lang.github.io/api/jason/infra/local/RunLocalMAS.html) you can get the [Architecture ](https://jason-lang.github.io/api/jason/infra/local/BaseLocalMAS.html#getAg(java.lang.String)) of running agents. From this class, you can get the Agent and then methods to add belifs.
3. you can take a look at the internal actions that make a snapshot of the agent to continue its execution later. [here](https://github.com/jason-lang/jason/tree/main/demos/snapshot).
4. If the class is not found, you likely have not set the classpath correctly.

**Reply by jomifred** [reply]
for 2. you can try

```
RunLocalMAS.getRunner().getAg("bob").getTS().getAg().addBel(.....)
```

**Reply by jomifred** [reply]
some UML diagrams from Jason book that may help

- [BB](https://github.com/user-attachments/files/17438784/cdBB.pdf)
- [Agent](https://github.com/user-attachments/files/17438810/cdAgent.pdf)
- [AgArch](https://github.com/user-attachments/files/17438825/cdAgArch.pdf)

they could be a little outdated, but the main ideas of the classes are there

**Reply by ChrisL20J** [reply]
Hello Jomi,
Thank you very much for answering my questions. 
Your answer was really helpful.
